File: src/CmdPlot.cpp
// CmdPlot.cpp - This file is part of QPlot

/* QPlot - Publication quality 2D graphs with dual coordinate systems
   Copyright (C) 2014  Daniel Wagenaar
  
   This program is free software: you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation, either version 3 of the License, or
   (at your option) any later version.
  
   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.
  
   You should have received a copy of the GNU General Public License
   along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

// CmdPlot.C

#include "CmdPlot.h"
#include "Rotate.h"
#include "Range.h"
#include <QDebug>
#include <cmath>

static CBuilder<CmdPlot> cbPlot("plot");
static CBuilder<CmdPlot> cbPatch("patch");
static CBuilder<CmdPlot> cbLine("line");
static CBuilder<CmdPlot> cbArea("area");

bool CmdPlot::usage() {
  return error("Usage: plot|patch|line|area xdata ydata");
}

bool CmdPlot::parse(Statement const &s) {
  int id1 = s.nextIndex(1);
  int id2 = s.nextIndex(id1);
  if (id2==s.length() && s.isNumeric(1) && s.isNumeric(id1) &&
      s.data(1).size()==s.data(id1).size())
    return true;
  else
    return usage();
}

QRectF CmdPlot::dataRange(Statement const &s) {
  if (s[0].str=="line" || s[0].str=="area")
    return QRectF();
  
  QVector<double> const &xdata = s.data(1);
  QVector<double> const &ydata = s.data(s.nextIndex(1));

  if (xdata.isEmpty())
    return QRectF();

  Range rangex, rangey;
  foreach (double x, xdata)
    rangex.extend(x);
  foreach (double y, ydata)
    rangey.extend(y);
  return QRectF(QPointF(rangex.min(),rangey.min()),
                QPointF(rangex.max(),rangey.max()));
}

void CmdPlot::render(Statement const &s, Figure &f, bool dryrun) {
  QVector<double> const &xdata = s.data(1);
  QVector<double> const &ydata = s.data(s.nextIndex(1));
  if (xdata.isEmpty()) {
    f.setBBox(QRectF());
    return;
  }

  QPainterPath p;
  bool lastnan = true;
  
  if (s[0].str=="plot" || s[0].str=="patch") {
    for (int k=0; k<xdata.size(); k++) {
      if (std::isnan(xdata[k]) || std::isnan(ydata[k])) {
        lastnan = true;
      } else {
        if (lastnan)
          p.moveTo(f.map(xdata[k],ydata[k]));
        else
          p.lineTo(f.map(xdata[k],ydata[k]));
        lastnan = false;
      }
    }
  } else {
    double a = f.anchorAngle();
    QPointF xy0 = f.anchor();
    for (int k=0; k<xdata.size(); k++) {
      if (std::isnan(xdata[k]) || std::isnan(ydata[k])) {
        lastnan = true;
      } else {
        QPointF xy(pt2iu(xdata[k]), pt2iu(ydata[k]));
        if (a)
          xy = ::rotate(xy, a);
        xy += xy0;
        if (lastnan)
          p.moveTo(xy);
        else
          p.lineTo(xy);
        lastnan = false;
      }
    }
  }

  if (s[0].str=="patch" || s[0].str=="area")
    p.closeSubpath();
  
  QRectF bbox = p.boundingRect();
  double w = f.painter().pen().widthF();
  if (w>0)
    bbox.adjust(-w/2, -w/2, w/2, w/2);
  /* Adjust for line width.
     This does not take protruding miters into account.
  */
  f.setBBox(bbox);

  if (dryrun)
    return;

  QBrush b = f.painter().brush();
  if (s[0].str=="plot" || s[0].str=="line")
    f.painter().setBrush(QBrush(Qt::NoBrush));

  f.painter().drawPath(p);

  if (s[0].str=="plot" || s[0].str=="line")
    f.painter().setBrush(b);
}

File: python/qplot_backend/cmdplot.py
# cmdplot.py - This file is part of QPlot

# QPlot - Publication quality 2D graphs with dual coordinate systems
# Copyright (C) 2014  Daniel Wagenaar
# (same GPL licence as original)

from __future__ import annotations

import math
from typing import TYPE_CHECKING

from PyQt6.QtCore import Qt, QPointF, QRectF
from PyQt6.QtGui import QBrush, QPainterPath

from .command import Command
from .token_ import Token
from .factor import pt2iu
from .range_ import Range
from .rotate import rotate_point

if TYPE_CHECKING:
    from .statement import Statement
    from .figure import Figure


@Command.register("plot")
@Command.register("patch")
@Command.register("line")
@Command.register("area")
class CmdPlot(Command):
    """Draw a line, filled polygon, or paper-space path.

    Syntax:
        plot  xdata ydata    — data-space polyline
        patch xdata ydata    — data-space closed polygon
        line  xdata ydata    — paper-space polyline (relative to anchor)
        area  xdata ydata    — paper-space closed polygon
    """

    def _usage(self) -> bool:
        return self.error("Usage: plot|patch|line|area xdata ydata")

    def parse(self, s: Statement) -> bool:
        id1 = s.next_index(1)
        id2 = s.next_index(id1)
        if (id2 == len(s)
                and s.is_numeric(1) and s.is_numeric(id1)
                and len(s.data(1)) == len(s.data(id1))):
            return True
        return self._usage()

    def data_range(self, s: Statement) -> QRectF:
        if s[0].str in ("line", "area"):
            return QRectF()  # paper-space commands have no data range

        xdata = s.data(1)
        ydata = s.data(s.next_index(1))
        if not xdata:
            return QRectF()

        range_x = Range()
        range_y = Range()
        for x in xdata:
            range_x.extend(x)
        for y in ydata:
            range_y.extend(y)
        return QRectF(
            QPointF(range_x.min(), range_y.min()),
            QPointF(range_x.max(), range_y.max()),
        )

    def render(self, s: Statement, f: Figure, dryrun: bool) -> None:
        xdata = s.data(1)
        ydata = s.data(s.next_index(1))

        if not xdata:
            f.set_bbox(QRectF())
            return

        path = QPainterPath()
        last_nan = True
        cmd = s[0].str

        if cmd in ("plot", "patch"):
            for k in range(len(xdata)):
                if math.isnan(xdata[k]) or math.isnan(ydata[k]):
                    last_nan = True
                else:
                    pt = f.map(xdata[k], ydata[k])
                    if last_nan:
                        path.moveTo(pt)
                    else:
                        path.lineTo(pt)
                    last_nan = False

        else:  # line / area — paper space, relative to anchor
            a = f.anchor_angle()
            xy0 = f.anchor()
            for k in range(len(xdata)):
                if math.isnan(xdata[k]) or math.isnan(ydata[k]):
                    last_nan = True
                else:
                    xy = QPointF(pt2iu(xdata[k]), pt2iu(ydata[k]))
                    if a:
                        xy = rotate_point(xy, a)
                    xy += xy0
                    if last_nan:
                        path.moveTo(xy)
                    else:
                        path.lineTo(xy)
                    last_nan = False

        if cmd in ("patch", "area"):
            path.closeSubpath()

        bbox = path.boundingRect()
        w = f.painter().pen().widthF()
        if w > 0:
            bbox.adjust(-w/2, -w/2, w/2, w/2)
        # Note: does not account for protruding miters.
        f.set_bbox(bbox)

        if dryrun:
            return

        brush = f.painter().brush()
        if cmd in ("plot", "line"):
            f.painter().setBrush(QBrush(Qt.BrushStyle.NoBrush))

        f.painter().drawPath(path)

        if cmd in ("plot", "line"):
            f.painter().setBrush(brush)
